runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,239 @@
+      hints_by_lnum[lnum] = nil
+    end
+  end
+
+  local countdown = function()
+    num_unprocessed = num_unprocessed - 1
+    if num_unprocessed == 0 then
+      bufstate.version = version
+      vim.cmd('redraw!')
+    end
+  end
+
+  local redraw = false
+  for lnum, new_line_hints in pairs(new_hints_by_lnum) do
+    local num_unresolved = 0
+    for _, hint in pairs(new_line_hints) do
+      if not hint._resolved then
+        num_unresolved = num_unresolved + 1
+      end
+    end
+    local current_line_hints = hints_by_lnum[lnum]
+    if not next(current_line_hints or {}) then
+      hints_by_lnum[lnum] = new_line_hints
+      redraw = true
+    end
+    for _, hint in pairs(new_line_hints) do
+      if hint._resolved then
+        countdown()
+      else
+        client.request('inlayHint/resolve', hint, function(err, result)
+          if err then
+            hint._resolved = false
+            return
+          end
+          if result then
+            hint.kind = result.kind and result.kind or hint.kind
+            hint.textEdits = result.textEdits and result.textEdits or hint.textEdits
+            hint.label = result.label and result.label or hint.label
+          end
+          hint._resolved = true
+          num_unresolved = num_unresolved - 1
+          if num_unresolved == 0 and current_line_hints then
+            hints_by_lnum[lnum] = new_line_hints
+          end
+          countdown()
+        end)
+      end
+    end
+  end
+  if redraw then
+    vim.cmd('redraw!')
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/inlayHint`
+---
+function M.on_inlayhint(err, result, ctx, _)
+  if err then
+    local _ = log.error() and log.error('inlayhint', err)
+    return
+  end
+  M.save(result, ctx.bufnr, ctx.client_id, ctx.version)
+end
+
+---@private
+local function resolve_bufnr(bufnr)
+  return bufnr == 0 and api.nvim_get_current_buf() or bufnr
+end
+
+--- Refresh the inlay hints for the current buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---
+--- Example:
+--- <pre>vim
+---   autocmd BufEnter,InsertLeave,BufWritePost <buffer> lua vim.lsp.inlay_hint.refresh()
+--- </pre>
+---
+function M.refresh(bufnr)
+  bufnr = resolve_bufnr(bufnr or 0)
+  local params = {
+    textDocument = util.make_text_document_params(bufnr),
+    range = {
+      start = { line = 0, character = 0 },
+      ['end'] = { line = api.nvim_buf_line_count(bufnr), character = 0 },
+    },
+  }
+  vim.lsp.buf_request(bufnr, 'textDocument/inlayHint', params)
+end
+
+--- |lsp-handler| for the method `workspace/inlayHint/refresh`
+---@see https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_inlayHint_refresh
+handlers['workspace/inlayHint/refresh'] = function(err, _, ctx)
+  if err then
+    return vim.NIL
+  end
+
+  for _, bufnr in ipairs(vim.lsp.get_buffers_by_client_id(ctx.client_id)) do
+    M.refresh(bufnr)
+  end
+
+  return vim.NIL
+end
```

Comment:
Could you move this to the `handlers.lua` ? Doing this in a separate module breaks the existing pattern (and if the imports weren't lazy, it would cause cyclic issues)

And I think it may be worth checking if the buffers are active in any window. Otherwise this could have a big performance impact if you've a lot of buffers open and it immediately refreshes all of them incluing resolve.

Reply:
> Could you move this to the handlers.lua ? Doing this in a separate module breaks the existing pattern (and if the imports weren't lazy, it would cause cyclic issues)

semantic tokens does this and none of the refresh ones are in handlers.lua so i did this, I'll move it 👍🏽 

> And I think it may be worth checking if the buffers are active in any window. Otherwise this could have a big performance impact if you've a lot of buffers open and it immediately refreshes all of them incluing resolve.

will do